fiona/_geometry.pyx

```diff
@@ -143,20 +146,49 @@ cdef class GeomBuilder:
 
     cdef build(self, void *geom):
         # The only method anyone needs to call
-        if geom == NULL:
-            raise ValueError("Null geom")
-
-        cdef unsigned int etype = OGR_G_GetGeometryType(geom)
+        cdef void *cogr_geometry = NULL
+        cdef void *cloned_geometry = NULL
 
-        self.code = base_geometry_type_code(etype)
+        if geom == NULL:
+            return None
+
+        code = base_geometry_type_code(OGR_G_GetGeometryType(geom))
+
+        # We convert special geometries (curves, TIN, Triangle, ...)
+        # to GeoJSON compatible geometries (LineStrings, Polygons, MultiPolygon, ...)
+        if 8 <= code <= 14:  # Curves.
+            # OGR_G_GetLinearGeometry: The ownership of the returned geometry belongs to the caller.
+            cogr_geometry = OGR_G_GetLinearGeometry(geom, 0.0, NULL)
+            code = base_geometry_type_code(OGR_G_GetGeometryType(cogr_geometry))
+            ndims = OGR_G_GetCoordinateDimension(cogr_geometry)
+            geom = cogr_geometry
+        elif 15 <= code <= 17:
+            cloned_geometry = OGR_G_Clone(geom)
+            if code in (15, 16):
+                # OGR_G_ForceToMultiPolygon: the converted geometry (ownership to caller).
+                cogr_geometry = OGR_G_ForceToMultiPolygon(cloned_geometry)
+            elif code == 17:
+                # OGR_G_ForceToPolygon: the converted geometry (ownership to caller).
+                cogr_geometry = OGR_G_ForceToPolygon(cloned_geometry)
+            code = base_geometry_type_code(OGR_G_GetGeometryType(cogr_geometry))
+            ndims = OGR_G_GetCoordinateDimension(cogr_geometry)
+            geom = cogr_geometry
+        else:
+            ndims = OGR_G_GetCoordinateDimension(geom)
 
-        if self.code not in GEOMETRY_TYPES:
-            raise UnsupportedGeometryTypeError(self.code)
+        if code not in GEOMETRY_TYPES:
+            raise UnsupportedGeometryTypeError(code)
 
-        self.geomtypename = GEOMETRY_TYPES[self.code]
-        self.ndims = OGR_G_GetCoordinateDimension(geom)
+        self.code = code
+        self.geomtypename = GEOMETRY_TYPES[code]
+        self.ndims = ndims
         self.geom = geom
         built = getattr(self, '_build' + self.geomtypename)()
+
+        # Cleanup geometries we have ownership over
+        if cogr_geometry is not NULL:
```

Comment:
It is probably not necessary, but it would be more reassuring to nullify self.geom as you are actually destroying it just after...
```suggestion
        if cogr_geometry is not NULL:
            self.geom = None
```